### PR TITLE
refactor(url-state-provider): refactor to typescript

### LINKED ISSUE #163: [TASK]: Convert @cloudoperators/juno-url-state-provider to TypeScript
### Task Description

We need to convert  @cloudoperators/juno-state-provider to TypeScript

**Sub-tasks**

- [ ] Integrate Lint with typescript config(look to juno-communicator package as example)
- [ ] Integrate tsconfig.json from config package
- [ ] Integrate vitest for testing
- [ ] Integrate tsup for bundling things up
- [ ] Delete old rollup/esbuild/babel
- [ ] Makes tests green integrate in example app.
- [ ] Update readme

**Related Issues**
- #119 

**Additional Context**
Look to already existing configuration in juno-communicator

### PR DESCRIPTION
**Pull request by TilmanHaupt**
# Summary

This pull request refactors the app url-state-provider and all associated tests from JavaScript to TypeScript. The transition aims to improve type safety, enhance code maintainability, and leverage TypeScript's features for better development experience.

# Related Issues

closes #163 

# Testing Instructions

1. `npm i`
2. `npm run TASK`

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

### CODE REVIEW

**Review by andypf — COMMENTED** [review]
From my perspective, it looks good overall. I’m a bit uncertain about the use of any types, but aside from that, everything else looks solid. Nice work!

File: packages/url-state-provider/src/index.ts [hunk 1]
```diff
@@ -112,8 +128,8 @@ function stateToURL(state) {
  * @param {string} stateID
  * @returns state for stateID
  */
-function currentState(stateID) {
-  return URLToState()[stateID]
+function currentState(stateID: string): object {
+  return URLToState()[stateID] as object
```
Comment (barsukov) [inline packages/url-state-provider/src/index.ts:132]: do you need to cast here as object? maybe you could specify a type here for this type of thing

File: packages/url-state-provider/src/index.ts [hunk 2]
```diff
@@ -122,17 +138,17 @@ function currentState(stateID) {
  * @param {JSON} state
  * @returns new url string
  */
-function updateState(stateID, state, options = {}) {
+function updateState(stateID: string, state: object, options: UpdateStateOptions = {}) {
   // get current state from URL
-  const newState = URLToState()
+  const newState: { [key: string]: any } = URLToState()
```
Comment (barsukov) [inline packages/url-state-provider/src/index.ts:143]: unknown

File: packages/url-state-provider/src/index.ts [hunk 3]
```diff
@@ -152,16 +168,15 @@ function updateState(stateID, state, options = {}) {
...
-function updateStateAndHistory(stateID, state, merge, historyOptions) {
-  historyOptions = historyOptions || {}
+function updateStateAndHistory(stateID: string, state: object, merge: boolean, historyOptions: HistoryOptions = {}) {
   const newUrl = updateState(stateID, state, { merge })
   const historyState = historyOptions.state || ""
   const historyTitle = historyOptions.title || ""
 
   if (historyOptions?.replace) {
-    window.history.replaceState(historyState, historyTitle, newUrl)
+    window.history.replaceState(historyState, historyTitle, newUrl as string)
```
Comment (barsukov) [inline packages/url-state-provider/src/index.ts:177]: avoid casting modify origin of method to return string

File: packages/url-state-provider/src/index.ts [hunk 4]
```diff
@@ -152,16 +168,15 @@ function updateState(stateID, state, options = {}) {
...
   } else {
-    window.history.pushState(historyState, historyTitle, newUrl)
+    window.history.pushState(historyState, historyTitle, newUrl as string)
```
Comment (barsukov) [inline packages/url-state-provider/src/index.ts:179]: here too

File: packages/url-state-provider/src/index.ts [hunk 5]
```diff
@@ -255,7 +273,7 @@ function informListener(stateID, newState, oldState) {
   listener(newState)
 }
 
-function onGlobalChange(callback) {
+function onGlobalChange(callback: any): any {
```
Comment (barsukov) [inline packages/url-state-provider/src/index.ts:276]: please specify type for callback

File: packages/url-state-provider/src/superstate.test.ts [hunk 6]
```diff
@@ -358,7 +359,7 @@ describe("broken LZ decompression", () => {
 
     try {
       humanURI.decodeLZ(data)
-    } catch (error) {
+    } catch (error: any) {
```
Comment (barsukov) [inline packages/url-state-provider/src/superstate.test.ts:362]: error type?

File: packages/url-state-provider/src/superstate.test.ts [hunk 7]
```diff
@@ -375,7 +376,7 @@ describe("broken B64 decomopression", () => {
 
     try {
       humanURI.decodeB64(data)
-    } catch (error) {
+    } catch (error: any) {
```
Comment (barsukov) [inline packages/url-state-provider/src/superstate.test.ts:379]: error type?

File: packages/url-state-provider/src/superstate.test.ts [hunk 8]
```diff
@@ -392,7 +393,7 @@ describe("broken decoding because of not closed obj", () => {
 
     try {
       humanURI.decode(data)
-    } catch (error) {
+    } catch (error: any) {
```
Comment (barsukov) [inline packages/url-state-provider/src/superstate.test.ts:396]: error type?

File: packages/url-state-provider/src/superstate.test.ts [hunk 9]
```diff
@@ -404,7 +405,7 @@ describe("broken decoding because of not closed obj", () => {
     const data = "%(324"
     try {
       humanURI.decode(data)
-    } catch (error) {
+    } catch (error: any) {
```
Comment (barsukov) [inline packages/url-state-provider/src/superstate.test.ts:408]: error type?

**Review by barsukov — CHANGES_REQUESTED** [review]
Please have a look to the comments nice start :)

File: packages/url-state-provider/src/index.ts (hunk 1 above)
Comment (TilmanHaupt) [reply packages/url-state-provider/src/index.ts:132]: I don't fully understand but thats was linter is saying.

File: packages/url-state-provider/src/index.ts (hunk 2 above)
Comment (TilmanHaupt) [reply packages/url-state-provider/src/index.ts:143]: if i do that there is the error "spread can only be created from object types.

File: packages/url-state-provider/src/index.ts (hunk 3 above)
Comment (TilmanHaupt) [reply packages/url-state-provider/src/index.ts:177]: From how I understand andreas methods it is, in fact a object. Which is somehow saved as a string.

File: packages/url-state-provider/src/index.ts (hunk 4 above)
Comment (TilmanHaupt) [reply packages/url-state-provider/src/index.ts:179]: look other comment.

**Review by barsukov — APPROVED** [review]
Please create @TilmanHaupt leftover ticket for aslant any issues that we spoke about. Other then that it is a really good job 🥇

**Review by andypf — APPROVED** [review]

**Pull request merged**